src/main/java/bio/terra/service/resourcemanagement/flight/RecordBucketAutoclassStep.java

```diff
@@ -0,0 +1,57 @@
+package bio.terra.service.resourcemanagement.flight;
+
+import bio.terra.service.filedata.flight.FileMapKeys;
+import bio.terra.service.resourcemanagement.exception.GoogleResourceException;
+import bio.terra.service.resourcemanagement.google.GoogleBucketResource;
+import bio.terra.service.resourcemanagement.google.GoogleBucketService;
+import bio.terra.stairway.FlightContext;
+import bio.terra.stairway.FlightMap;
+import bio.terra.stairway.Step;
+import bio.terra.stairway.StepResult;
+import bio.terra.stairway.StepStatus;
+import com.google.cloud.storage.Bucket;
+import com.google.cloud.storage.BucketInfo.Autoclass;
+import com.google.cloud.storage.StorageClass;
+
+public class RecordBucketAutoclassStep implements Step {
+
+  private final GoogleBucketService googleBucketService;
+  private final String bucketName;
+
+  public RecordBucketAutoclassStep(GoogleBucketService googleBucketService, String bucketName) {
+    this.googleBucketService = googleBucketService;
+    this.bucketName = bucketName;
+  }
+
+  private StepResult stepResultFailure(String message) {
+    return new StepResult(
+        StepStatus.STEP_RESULT_FAILURE_FATAL, new GoogleResourceException(message));
+  }
+
+  @Override
+  public StepResult doStep(FlightContext context) throws InterruptedException {
+    FlightMap workingMap = context.getWorkingMap();
+    Bucket bucket = googleBucketService.getCloudBucket(bucketName);
+    GoogleBucketResource bucketResource = googleBucketService.getBucketMetadata(bucketName);
+    if (bucket == null || bucketResource == null) {
+      return stepResultFailure("Bucket not found: " + bucketName);
+    }
+    Autoclass autoclass = bucket.getAutoclass();
+    if (autoclass.getEnabled() && autoclass.getTerminalStorageClass() == StorageClass.ARCHIVE) {
+      return stepResultFailure("Bucket autoclass already set to ARCHIVE: " + bucketName);
+    }
+    if (autoclass.getEnabled() ^ bucketResource.getAutoclassEnabled()) {
+      return stepResultFailure("Bucket autoclass mismatch in metadata: " + bucketName);
+    }
+    if (autoclass.getEnabled()) {
```

Comment:
autoclass getEnabled() could be a variable maybe. Also, can you explain this step again, we want to record the current state in case of a rollback right? why fail if the bucket is already set to archive? why have a separate case for if autoclass is enabled compared to if it isn't - what is being recorded differently?

Reply:
We don't _have_ to fail if the bucket is Autoclass is set to ARCHIVE, but if we don't we end up applying the same policies again to the bucket, so I prefer to fail fast instead.

As for your other question, the problem is that Google changed how the Autoclass behavior works in October of 2023. Before it was the case that Autoclass terminated at ARCHIVE automatically, but now Autoclass terminates at NEARLINE instead. So it's not enough to simply rollback the Autoclass state, but the terminal storage class must also be recorded to be able to rollback as well.

In other words, we need to store the current class of the bucket, whether Autoclass is turned on, and what the Autoclass terminal class is. All three of these must be restored to their previous values in case of rollback.